proposals/webidl-bindings/bindings.md

```diff
@@ -0,0 +1,508 @@
+WebAssembly code can be interpreted in terms of implementing a schema defined
+API.
+
++ Principal Function Call
+
+When constructing an access lambda, or when constructing a provision lambda
+there is often a function call within the appropriate binding expression that
+represents the main or principal call. In the case of an access lambda, this
+principal call is normally a call to a wi-IDL function; whereas in the case of
+a provision lambda, the principal function call is typically to a webAssembly
+function that has been exported.
+
+However, for a variety of reasons, there may be multiple calls expressed in an
+acess or provision lambda. For example, any time a `string` value is
+communicated (either as the result of a principal call or as an argument to a
+principal call) there may be additional calls to functions that allocate the
+string value in linear memory and/or calls to functions that lift a block of
+linear memory into a string value.
+
+From a technical perspective, there is actually very little that distinguishes
+such ancillary calls from the principal function call. Furthermore, there may be
+use cases (such as when testing a webAssembly module) where there is no actual
+principal function invoked.
+
+
+## Abstract Syntax
+
+This section defines the legal forms of interoperability bindings. In parallel
+we also show the 'W-expression' form of each term.
+
+>There are many cases where it is important to understand whether a given
+>expression refers to an IDL concept or whether it refers to a WebAssembly
+>concept. As an aid to clarity, we mark identifiers that refer to a WebAssembly
+>concept with a prefix `$`. Identifiers that refer to IDL concepts are not
+>marked; and identifiers that refer to both WebAssembly and to IDL concepts (such
+>as binding operators) will be marked if the _result_ of the operator refers to a
+>WebAssembly element.
+
+### Binding Lambda
+
+A _BindingLambda_ is a function that defines how a wasm function may access or
+provide an external API function.
+
+```
+BindingLambda => AccessLambda
+BindingLambda => ProvisionLambda
+```
+
+### Access Lambda
+
+An _AccessLambda_ is a function that defines how an imported function should be
+realized as an EB Function.
+
+```
+AccessLambda => 'access' Tuple '->' AccessExpression
+```
+
+```
+W-AccessLambda ::= '(' 'access' W-Tuple W-AccessExpression ')'
+```
+
+For example, the access lambda 
+
+
+#### Access Lambda Type
+
+An access lambda has a type signature that is expressed as a WASM function type.
+
+```
+E |= AccessLambda : functype
+```
+
+### Provision Lambda
+
+A _ProvisionLambda_ is a function that defines how an exported EB function is
+realized term terms of exported wasm elements.
+
+```
+ProvisionLambda => 'provide' Tuple '=>' ProvisionExpression
+```
+
+or, as a W-Expression form:
+
+```
+W-ProvisionLambda ::= '(' 'provide' W-Tuple W-ProvisionExpression ')'
+```
+
+#### Provision Lambda Type
```

Comment:
So the idea is that instead of

* mapping tuples of Web IDL values into tuples of Wasm values (incoming binding maps in the explainer) and vice versa (outgoing binding maps in the explainer),
* and then combining an incoming and outgoing binding map into either an import binding (with an outgoing binding map for arguments and incoming binding map for return values) or an export binding (with an incoming binding map for arguments and outgoing binding map for return values)

we now have

* access lambdas and provision lambdas that contain both lifting and lowering operators,
* but the lifting/lowering operators are separated from each other by which side (parameter or return) of the function whose values they are coercing
* and we have `^` to splice nested lists into the current list

correct?

The original approach has two benefits, as I see it:

1. The incoming and outgoing binding maps contain only lowering and lifting coercions respectively, while the access and provision lambdas contain heterogeneous expression trees of coercions. The lack of heterogeneity in the original allows for reusing opcodes across lifting and lowering coercions, which means that ultimately we can keep all the opcodes in the one-byte LEB128 space longer before reaching any two-byte LEB128s, which helps code size.

2. The original incoming/outgoing binding map formulation didn't require list splicing operators. To convert into `n` values, you just provided `n` expressions and each operator evaluated to a single value. The new approach requires an "hourglass" shape and the `^` splicer since the function being bound exists as a single W-expression but can return many values.